### ISSUE TITLE
KeyError edge case with `itertools.product` solution

### ISSUE THREAD

**Issue by diceroll123** [sourcery-ai/sourcery#215]
### Issue description or question

In some cases, itertools.product is unfortunately not the right way to go.

Consider:

```python
result = {'blah': 0, 'hello': 1}
for _a, _b in zip(range(2), range(2)):
    for key in list(result.keys()):
        # pretend I had logic here that
        # only sometimes made the next line happen
        result.pop(key)
```

This works fine, but the `itertools.product` solution changes the big O notation, so the inner function runs more than it can with the same value.

```python
import itertools
result = {'blah': 0, 'hello': 1}
for (_a, _b), key in itertools.product(zip(range(2), range(2)), list(result.keys())):
    # pretend I had logic here that
    # only sometimes made the next line happen
    result.pop(key)
```

Error returned: 
```
Traceback (most recent call last):
  line 6, in <module>
KeyError: 'blah'
```

_My example code is a very slimmed down version of the code I actually encountered this bug in, yes I know it's a travesty thank you._ 😂 

Thanks for readin'.

### Sourcery Version

0.11.1

### Code editor or IDE name and version

Version: 1.65.2 (user setup)
Commit: c722ca6c7eed3d7987c0d5c3df5c45f6b15e77d1
Date: 2022-03-10T14:33:55.248Z
Electron: 13.5.2
Chromium: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Windows_NT x64 10.0.22581

**Comment by brendanator** [comment]
That's a very interesting example. Thanks for reporting, I will fix alongside the other `itertools.product` issue you raise

**Comment by brendanator** [comment]
I've identified a fix and it will be included in the next release

**Comment by Hellebore** [comment]
Fixed in 0.11.3
